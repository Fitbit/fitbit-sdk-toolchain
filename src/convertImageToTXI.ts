import { Transform, Readable } from 'stream';

import {
  encode,
  TXIOutputFormat,
  RGBAOutputFormat,
} from '@fitbit/image-codec-txi';
import PluginError from 'plugin-error';
import { Metadata as PNGMetadata, PNG } from '@fitbit/pngjs';
import Vinyl from 'vinyl';

export { TXIOutputFormat };

const PLUGIN_NAME = 'convertImageToTXI';

export interface ConvertImageToTXIOptions {
  /**
   * Defaults to RGBA8888
   */
  rgbaOutputFormat?: RGBAOutputFormat;
}

function pickOutputFormat(
  image: PNGMetadata,
  rgbaOutputFormat: RGBAOutputFormat = TXIOutputFormat.RGBA8888,
) {
  if (!image.color) return TXIOutputFormat.A8;
  if (!image.alpha) return TXIOutputFormat.RGB565;

  return rgbaOutputFormat;
}

function transformPNG(png: PNG, rgbaOutputFormat?: RGBAOutputFormat) {
  return new Promise<Buffer>((resolve, reject) => {
    png.on('metadata', (metadata) => {
      png.on('parsed', () => {
        try {
          const txi = encode(
            {
              data: new Uint8ClampedArray(png.data),
              width: png.width,
              height: png.height,
            },
            {
              outputFormat: pickOutputFormat(metadata, rgbaOutputFormat),
              rle: 'auto',
            },
          );
          resolve(Buffer.from(txi));
        } catch (e) {
          reject(e);
        }
      });
    });
    png.on('error', reject);
  });
}

function transformPNGBuffer(file: Buffer, rgbaOutputFormat?: RGBAOutputFormat) {
  const png = new PNG();
  const transformer = transformPNG(png, rgbaOutputFormat);
  png.parse(file);
  return transformer;
}

/**
 * Lazily parse the PNG to cut down on the number of images
 * concurrently in memory.
 */
function transformPNGStream(
  file: NodeJS.ReadableStream,
  rgbaOutputFormat?: RGBAOutputFormat,
) {
  let started = false;

  return new Readable({
    read() {
      if (started) return;
      started = true;

      const png = new PNG();
      transformPNG(png, rgbaOutputFormat)
        .then((txi) => {
          this.push(txi);
          this.push(null);
        })
        .catch((error) => this.emit('error', error));

      file.on('error', (err) => this.emit('error', err)).pipe(png);
    },
  });
}

export default function convertImageToTXI(
  options: ConvertImageToTXIOptions = {},
) {
  return new Transform({
    objectMode: true,
    transform(this: Transform, file: Vinyl, _, cb) {
      if (file.isNull() || file.extname !== '.png') {
        return cb(undefined, file);
      }

      const fileName = file.relative;
      file.basename += '.txi';

      if (file.isBuffer()) {
        transformPNGBuffer(file.contents, options.rgbaOutputFormat)
          .then((txi) => {
            file.contents = txi;
            cb(undefined, file);
          })
          .catch((err) => cb(new PluginError(PLUGIN_NAME, err, { fileName })));
      } else if (file.isStream()) {
        file.contents = transformPNGStream(
          file.contents,
          options.rgbaOutputFormat,
        ).on('error', (err) =>
          this.emit('error', new PluginError(PLUGIN_NAME, err, { fileName })),
        );
        cb(undefined, file);
      }
    },
  });
}
