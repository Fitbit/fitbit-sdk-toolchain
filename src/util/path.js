export function dirname(path) {
  let dir = path.split('/').slice(0, -1).join('/');
  if (dir === '') dir = '/';
  return dir;
}

export function basename(path) {
  return path.replace(/\\/g, '/').replace(/.*\//, '');
}

export function split(path) {
  return [dirname(path), basename(path)];
}
