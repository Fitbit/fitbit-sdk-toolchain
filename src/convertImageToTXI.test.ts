import concatStream from 'concat-stream';
import Vinyl from 'vinyl';
import vinylFS from 'vinyl-fs';

import convertImageToTXI, { TXIOutputFormat } from './convertImageToTXI';
import { Readable, PassThrough } from 'stream';
import { join } from 'path';
import { readFileSync } from 'fs';

const corruptImage = join(
  __dirname,
  '__test__',
  'PngSuite-2017jul19',
  'xs1n0g01.png',
);

function eventCounter(total: number, callback: () => void) {
  let counter = total;
  if (counter < 1) throw new TypeError('positive number required');

  return () => {
    counter -= 1;
    if (counter < 1) callback();
  };
}

it('passes through non-PNG files untouched', (done) => {
  const files = [
    new Vinyl({
      path: 'index.js',
      contents: Buffer.of(),
    }),
    new Vinyl({
      path: 'directory.png',
      stat: {
        isDirectory: () => true,
      } as any,
    }),
  ];

  const handleData = jest.fn();

  const converter = convertImageToTXI();
  converter
    .on('error', done)
    .on('data', handleData)
    .on('end', () => {
      expect(handleData).toHaveBeenCalledTimes(files.length);
      files.forEach((value, index) =>
        expect(handleData).toHaveBeenNthCalledWith(index + 1, value),
      );
      done();
    });

  files.forEach((file) => converter.write(file));
  converter.end();
});

describe('in streaming mode', () => {
  it('does not start reading from the source stream until there is a sink', (done) => {
    let shouldRead = false;
    const fileStream = new Readable({
      read() {
        if (shouldRead) {
          done();
        } else {
          done('Started reading too soon');
        }
      },
    });

    convertImageToTXI()
      .on('data', (file: Vinyl) =>
        process.nextTick(() => {
          shouldRead = true;
          if (!file.isStream()) {
            done('Got a non-stream file');
            return;
          }
          file.contents.on('data', () => {});
        }),
      )
      .end(
        new Vinyl({
          path: 'foo.png',
          contents: fileStream,
        }),
      );
  });

  it('handles a stream read error', (done) => {
    expect.assertions(2);

    const fileStream = new Readable({
      read() {
        process.nextTick(() => this.emit('error', new Error('blarg im ded')));
      },
    });

    const converter = convertImageToTXI();
    converter
      .on('data', (file: Vinyl) => {
        if (!file.isStream()) {
          done('Got a non-stream file');
          return;
        }

        // Start the lazy transform.
        file.contents
          .on('error', (error) => expect(error).toMatchSnapshot())
          .on('data', () => {});
      })
      .on('error', (error) => expect(error).toMatchSnapshot())
      .on('end', done);

    converter.end(
      new Vinyl({
        path: 'dead.png',
        contents: fileStream,
      }),
    );
  });

  it('handles a corrupt PNG', (done) => {
    expect.assertions(2);
    const tally = eventCounter(2, done);

    vinylFS
      .src(corruptImage, { buffer: false })
      .pipe(convertImageToTXI())
      .on('error', (error) => {
        expect(error).toMatchSnapshot();
        tally();
      })
      .on('data', (file: Vinyl) => {
        if (!file.isStream()) {
          done('Got a non-stream file');
          return;
        }

        file.contents
          .on('error', (error) => {
            expect(error).toMatchSnapshot();
            tally();
          })
          .on('data', () => {});
      });
  });

  // FIXME: Jest typings are broken and can't deal with the done callback parameter properly
  // https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34617
  it.each<any>([
    ['empty', Buffer.of()],
    ['truncated', Buffer.from('89504E470D0A1A0A', 'hex')],
  ])('gracefully errors if the PNG file is %s', (_, data, done) => {
    expect.assertions(2);
    const tally = eventCounter(2, done);

    const contents = new PassThrough();
    contents.end(data);

    convertImageToTXI()
      .on('error', (error) => {
        expect(error).toMatchSnapshot();
        tally();
      })
      .on('data', (file: Vinyl) => {
        if (!file.isStream()) {
          done('Got a non-stream file');
          return;
        }

        file.contents
          .on('error', (error) => {
            expect(error).toMatchSnapshot();
            tally();
          })
          .on('data', () => {});
      })
      .end(
        new Vinyl({
          contents,
          path: 'image.png',
        }),
      );
  });

  // FIXME: Jest typings are broken and can't deal with the done callback parameter properly
  // https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34617
  it.each<any>([
    ['1bit.png', 'monochrome'],
    ['rgb_image.png', 'RGB'],
    ['rle_no_leftovers.png', 'RGBA'],
  ])(
    'converts %s to a %s TXI',
    (filename: string, _, done: jest.DoneCallback) => {
      expect.hasAssertions();

      const png = join(__dirname, '__test__', filename);
      vinylFS
        .src(png, { buffer: false })
        .pipe(convertImageToTXI())
        .on('error', done)
        .on('data', (txi: Vinyl) => {
          expect(txi.basename).toBe(`${filename}.txi`);
          expect(txi.isStream()).toBe(true);
          if (txi.isStream()) {
            txi.contents.pipe(
              concatStream((contents) => {
                expect(contents.compare(readFileSync(`${png}.txi`))).toBe(0);
                done();
              }),
            );
          }
        });
    },
  );
});

describe('in buffered mode', () => {
  it('handles a corrupt PNG', (done) => {
    expect.assertions(1);

    vinylFS
      .src(corruptImage)
      .pipe(convertImageToTXI())
      .on('error', (error) => {
        expect(error).toMatchSnapshot();
        done();
      })
      .on('data', () => done('Got an unexpected file'));
  });

  // FIXME: Jest typings are broken and can't deal with the done callback parameter properly
  // https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34617
  it.each<any>([
    ['empty', Buffer.of()],
    ['truncated', Buffer.from('89504E470D0A1A0A', 'hex')],
  ])('gracefully errors if the PNG file is %s', (_, contents, done) => {
    expect.assertions(1);

    convertImageToTXI()
      .on('error', (error) => {
        expect(error).toMatchSnapshot();
        done();
      })
      .on('data', () => done('Got an unexpected file'))
      .end(
        new Vinyl({
          contents,
          path: 'image.png',
        }),
      );
  });

  // FIXME: Jest typings are broken and can't deal with the done callback parameter properly
  // https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34617
  it.each<any>([
    ['1bit.png', 'monochrome'],
    ['rgb_image.png', 'RGB'],
    ['rle_no_leftovers.png', 'RGBA'],
  ])(
    'converts %s to a %s TXI',
    (filename: string, _, done: jest.DoneCallback) => {
      expect.hasAssertions();

      const png = join(__dirname, '__test__', filename);
      vinylFS
        .src(png)
        .pipe(convertImageToTXI())
        .on('error', done)
        .on('data', (txi: Vinyl) => {
          expect(txi.basename).toBe(`${filename}.txi`);
          expect(txi.isBuffer()).toBe(true);
          if (txi.isBuffer()) {
            expect(txi.contents.compare(readFileSync(`${png}.txi`))).toBe(0);
            done();
          }
        });
    },
  );
});

describe('RGBA output format for alpha channel images', () => {
  it('converts to RGBA6666 if specified', (done: jest.DoneCallback) => {
    const png = join(__dirname, '__test__/rle_no_leftovers.png');

    vinylFS
      .src(png)
      .pipe(convertImageToTXI({ rgbaOutputFormat: TXIOutputFormat.RGBA6666 }))
      .on('error', done)
      .on('data', (txi: Vinyl) => {
        if (txi.isBuffer()) {
          expect(
            txi.contents.compare(readFileSync(`${png}.RGBA6666.txi`)),
          ).toBe(0);
          done();
        }
      });
  });

  it('defaults to RGBA8888 if RGBA6666 is not specified', (done: jest.DoneCallback) => {
    const png = join(__dirname, '__test__/rle_no_leftovers.png');

    vinylFS
      .src(png)
      .pipe(convertImageToTXI())
      .on('error', done)
      .on('data', (txi: Vinyl) => {
        if (txi.isBuffer()) {
          expect(txi.contents.compare(readFileSync(`${png}.txi`))).toBe(0);
          done();
        }
      });
  });
});
