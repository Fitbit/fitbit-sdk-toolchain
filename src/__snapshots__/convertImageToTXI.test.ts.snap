// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`in buffered mode gracefully errors if the PNG file is empty 1`] = `
PluginError {
  "fileName": "image.png",
  "message": "Unexpected end of input",
  "name": "Error",
  "plugin": "convertImageToTXI",
}
`;

exports[`in buffered mode gracefully errors if the PNG file is truncated 1`] = `
PluginError {
  "fileName": "image.png",
  "message": "Unexpected end of input",
  "name": "Error",
  "plugin": "convertImageToTXI",
}
`;

exports[`in buffered mode handles a corrupt PNG 1`] = `
PluginError {
  "fileName": "xs1n0g01.png",
  "message": "Invalid file signature",
  "name": "Error",
  "plugin": "convertImageToTXI",
}
`;

exports[`in streaming mode gracefully errors if the PNG file is empty 1`] = `
PluginError {
  "fileName": "image.png",
  "message": "Unexpected end of input",
  "name": "Error",
  "plugin": "convertImageToTXI",
}
`;

exports[`in streaming mode gracefully errors if the PNG file is empty 2`] = `[Error: Unexpected end of input]`;

exports[`in streaming mode gracefully errors if the PNG file is truncated 1`] = `
PluginError {
  "fileName": "image.png",
  "message": "Unexpected end of input",
  "name": "Error",
  "plugin": "convertImageToTXI",
}
`;

exports[`in streaming mode gracefully errors if the PNG file is truncated 2`] = `[Error: Unexpected end of input]`;

exports[`in streaming mode handles a corrupt PNG 1`] = `
PluginError {
  "fileName": "xs1n0g01.png",
  "message": "Invalid file signature",
  "name": "Error",
  "plugin": "convertImageToTXI",
}
`;

exports[`in streaming mode handles a corrupt PNG 2`] = `[Error: Invalid file signature]`;

exports[`in streaming mode handles a stream read error 1`] = `
PluginError {
  "fileName": "dead.png",
  "message": "blarg im ded",
  "name": "Error",
  "plugin": "convertImageToTXI",
}
`;

exports[`in streaming mode handles a stream read error 2`] = `[Error: blarg im ded]`;
